src/app/pluginAdapter.js

```diff
@@ -29,3 +35,53 @@ export interface DynamicPluginAdapter {
   nodeDescription(NodeAddressT): string;
   static (): StaticPluginAdapter;
 }
+
+function findUniqueMatch<T>(
```

Comment:
Alternate implementation—a bit more declarative, but I don’t have strong
feelings either way:

```diff
diff --git a/src/app/pluginAdapter.js b/src/app/pluginAdapter.js
index d0bfa9a..f5e985a 100644
--- a/src/app/pluginAdapter.js
+++ b/src/app/pluginAdapter.js
@@ -7,6 +7,7 @@ import {
   type NodeAddressT,
   type EdgeAddressT,
 } from "../core/graph";
+import type {AddressModule} from "../core/address";
 import type {Repo} from "../core/repo";
 
 export type EdgeType = {|
@@ -36,16 +37,18 @@ export interface DynamicPluginAdapter {
   static (): StaticPluginAdapter;
 }
 
-function findUniqueMatch<T>(
-  xs: $ReadOnlyArray<T>,
-  predicate: (T) => boolean
+function dispatch<A, T>(
+  am: AddressModule<A>,
+  x: A,
+  adapters: $ReadOnlyArray<T>,
+  prefix: (T) => A
 ): T {
-  const results = xs.filter(predicate);
+  const results = adapters.filter((a) => am.hasPrefix(x, prefix(a)));
   if (results.length > 1) {
-    throw new Error("Multiple entities match predicate");
+    throw new Error("Multiple adapters with prefix matching " + am.toString(x));
   }
   if (results.length === 0) {
-    throw new Error("No entity matches predicate");
+    throw new Error("No adapters with prefix matching " + am.toString(x));
   }
   return results[0];
 }
@@ -54,34 +57,26 @@ export function staticDispatchByNode(
   adapters: $ReadOnlyArray<StaticPluginAdapter>,
   x: NodeAddressT
 ): StaticPluginAdapter {
-  return findUniqueMatch(adapters, (a) =>
-    NodeAddress.hasPrefix(x, a.nodePrefix())
-  );
+  return dispatch(NodeAddress, x, adapters, (a) => a.nodePrefix());
 }
 
 export function staticDispatchByEdge(
   adapters: $ReadOnlyArray<StaticPluginAdapter>,
   x: EdgeAddressT
 ): StaticPluginAdapter {
-  return findUniqueMatch(adapters, (a) =>
-    EdgeAddress.hasPrefix(x, a.edgePrefix())
-  );
+  return dispatch(EdgeAddress, x, adapters, (a) => a.edgePrefix());
 }
 
 export function dynamicDispatchByNode(
   adapters: $ReadOnlyArray<DynamicPluginAdapter>,
   x: NodeAddressT
 ): DynamicPluginAdapter {
-  return findUniqueMatch(adapters, (a) =>
-    NodeAddress.hasPrefix(x, a.static().nodePrefix())
-  );
+  return dispatch(NodeAddress, x, adapters, (a) => a.static().nodePrefix());
 }
 
 export function dynamicDispatchByEdge(
   adapters: $ReadOnlyArray<DynamicPluginAdapter>,
   x: EdgeAddressT
 ): DynamicPluginAdapter {
-  return findUniqueMatch(adapters, (a) =>
-    EdgeAddress.hasPrefix(x, a.static().edgePrefix())
-  );
+  return dispatch(EdgeAddress, x, adapters, (a) => a.static().edgePrefix());
 }
diff --git a/src/app/pluginAdapter.test.js b/src/app/pluginAdapter.test.js
index cdb1d07..87af06f 100644
--- a/src/app/pluginAdapter.test.js
+++ b/src/app/pluginAdapter.test.js
@@ -61,7 +61,7 @@ describe("app/pluginAdapter", () => {
         const {statics} = example();
         const zod = NodeAddress.fromParts(["zod"]);
         expect(() => staticDispatchByNode(statics, zod)).toThrowError(
-          "No entity matches"
+          "No adapters with prefix matching"
         );
       });
       it("errors if there are multiple matches", () => {
@@ -69,7 +69,7 @@ describe("app/pluginAdapter", () => {
         const statics = [staticFooAdapter, staticFooAdapter];
         const foo = NodeAddress.fromParts(["foo"]);
         expect(() => staticDispatchByNode(statics, foo)).toThrowError(
-          "Multiple entities match"
+          "Multiple adapters with prefix matching"
         );
       });
     });
```

Reply:
This patch seems more complicated to me.